Cmdline

konova8 opened the pull request:
Closes #15 
Closes #16

lucat1 committed:
initial cmdline tool impl

samuelemusiani committed:
Modify go.mod and go.sum

samuelemusiani committed:
Removed ugly file (made by me btw)

samuelemusiani committed:
SOLVE ME PLS

konova8 committed:
Working cmdline

konova8 committed:
`gofmt -w .` for formatting files

konova8 committed:
Working togglable recursion + kebab-case for dir with flags in CLI

lucat1 commented on check.go:
```diff
@@ -1,32 +1,67 @@
 
-func CheckDir(synta synta.Synta, fs fs.ReadDirFS, dirPath string) (err error) {
-    entries, err := fs.ReadDir(dirPath)
+func CustomReadDir(fsys fs.FS, name string) ([]fs.DirEntry, error) {
```
A questa hack interna diamogli un nome che inizia con la minuscola. Io la chiamerei `readDir` semplicemente.
Metterei un commento sopra del tipo:
```go
// readDir uses the `readDir` method if the filesystem implements `fs.ReadDirFS`, otherwise opens the path and parses it using the `ReadDirFile` interface. 
```

lucat1 commented on check.go:
```diff
@@ -1,32 +1,67 @@
+	}
+
+	list, err := dir.ReadDir(-1)
+	sort.Slice(list, func(i, j int) bool { return list[i].Name() < list[j].Name() })
```
La `readDir` nativa quindi fa sort in automatico? qua la emuliamo o cosa?

lucat1 replied:
Se la speranza era di controllare i file in ordine, questo non funziona sempre, perchè nel primo if ritorniamo brutalmente e questo sort non verrebbe applicato.
In ogni caso non vedo il punto di organizzare, non cambia veramente i nostri controlli, però se è il comportamento di base della `readDir` allora lascerei.

konova8 replied:
È il comportamento base di `readDiir`, ho copiato il sorgente proprio, lascerei così

lucat1 commented on check.go:
```diff
@@ -1,32 +1,67 @@
-			if err := CheckName(synta, file.Name(), true); err != nil {
-				dirPath = path.Join(dirPath, file.Name())
-				if err := CheckDir(synta, fs, dirPath); err != nil {
+			kebab_regexp := regexp.MustCompile("^[a-z0-9]+(-[a-z0-9]+)*$")
```
Mettiamolo in una costante fuori dalla funzione.

konova8 replied:
L'ho messa dentro una `var`, la costante non e' possibile visto che restituisce un puntatore (credo)

lucat1 commented on check.go:
```diff
@@ -1,32 +1,67 @@
+	return list, err
+}
+
+func CheckDir(synta synta.Synta, fs fs.FS, dirPath string, recursive bool, kebab_dir bool) (err error) {
```
rinominare `kebab_dir` in `kebabDir` che è la forma standard. magari potremo chiamarla in modo più esplicito tipo: `ensureKebabCasing` e fare quel controllo per tutti i file (oltre che per le cartelle)

lucat1 commented on check.go:
```diff
@@ -1,32 +1,67 @@
-				if err := CheckDir(synta, fs, dirPath); err != nil {
+			kebab_regexp := regexp.MustCompile("^[a-z0-9]+(-[a-z0-9]+)*$")
+			if kebab_dir && !kebab_regexp.Match([]byte(file.Name())) {
+				err = fmt.Errorf("Directories need to be in kebab-case")
```
Stampare anche il nome del file che è colpevole dell'errore mi raccomando!

lucat1 commented on cmd/linter.go:
```diff
@@ -0,0 +1,46 @@
+
+	err = filenameslinter.CheckDir(syntaFile, os.DirFS(pwd), flag.Arg(1), *recursive, *kebab_case_dir)
+	if err != nil {
+		fmt.Printf("Error checking directory: %v", err)
```
```suggestion
    extra := ""
    if recursive {
      extra = "recursively "
    }
		fmt.Printf("Error while %schecking directory:\n%v", extra, err)
```

lucat1 reviewed (changes requested):
Piccole cose

lucat1 commented:
A parte quello che ho scritto. Incorporerei `rs/zerolog` per fare logging ad ogni file che viene controllato.
Stamperei, sia in modalità debug/trace letteralmente ogni chiamata di `checkName`, ma anche in modalità info ogni file controllato se passa o meno. Così quando si riceve un errore è ancora più chiaro quale file abbia rotto (sarà l'ultimo appena stampato). Lo faccio io quando vengono sistemate queste due cose se vi va.

konova8 committed:
Add "recursively" to the error message

Co-authored-by: Luca Tagliavini <luca.tagliavini5@studio.unibo.it>

konova8 committed:
Review suggestion from lucat1

- Make `CustomReadDir` a private function
- Regexp for kebab-case as an external variable
- Check kebab-case for all files and directories

konova8 committed:
Fix regexp for kebab-case and fix error return

konova8 committed:
Add logging for each file with `rs/zerolog`

konova8 committed:
Fix naming convention + missing `\n`

konova8 committed:
Add test for `check.go`

konova8 committed:
Use pretty logging with CLI

konova8 commented on check.go:
```diff
@@ -51,7 +51,7 @@ func CheckDir(synta synta.Synta, fs fs.FS, dirPath string, recursive bool, ensur
 			return err
 		}
-		log.Info().Msg("Checking file/dir with name " + file.Name())
+		log.Info().Msg("Checking file/dir with name " + dirPath + "/" + file.Name())
```
Qui lo vogliamo lasciare? Perche' lo stampa in ogni caso (come con `go test`)

konova8 replied:
Potremmo usare il logging solo ed esclusivamente dentro a `linter.go` e passare queste informazioni in qualche modo, in realta' forse questi log possiamo proprio toglierli @lucat1 

Edit: Scherzavo, bisogna lasciarlo

lucat1 replied:
Risolto no?

konova8 committed:
Make CLI work also with `/` as last char of directory

konova8 committed:
Fix bug when checking folder with more than 2 subfolders

lucat1 commented on check.go:
```diff
@@ -1,32 +1,77 @@
+	return list, err
+}
+
+var kebabRegexp *regexp.Regexp = regexp.MustCompile("^[a-z0-9]+(-[a-z0-9]+)*?(.[a-z0-9]+)$")
```
Non capisco il `?`. `*` matcha già con 0 occorrenze.

lucat1 replied:
Penso che anche quello che segue sia sbagliato (il `.` senza escaping è "ogni carattere"). Magari meglio scrivere:
```
^[a-z0-9]+(-[a-z0-9]+)*(\.[a-z0-9]+)?$
```

lucat1 commented on cmd/linter.go:
```diff
@@ -0,0 +1,55 @@
+		os.Exit(5)
+	}
+
+	dirPath := path.Base(pwd + "/" + flag.Arg(1))
```
Assolutamen da usare `path.Join` e non la concatenazione `+ "/" +`.

lucat1 commented on check.go:
```diff
@@ -1,32 +1,77 @@
+			err = fmt.Errorf("Could not read directory: %v", err)
+			return err
+		}
+		log.Info().Msg("Checking file/dir with name " + dirPath + "/" + file.Name())
```
Usare `path.Join` e non la concatenazione.

lucat1 reviewed (changes requested):
Due piccolezze.

lucat1 committed:
simplify the logic; clarify the logs; fix scanning the pwd

lucat1 committed:
add (default) option to ignore dotfiles
